File: code-library/04-logistic-regression/logistic-regression.py
# ==============================================================================
#  MODULE 4: LOGISTIC REGRESSION
#  Dataset: CES 2020 (cleaned in Module 1)
#
#  Stack: statsmodels (logit/probit), sklearn (ROC/AUC), matplotlib
# ==============================================================================

import pandas as pd
import numpy as np
import statsmodels.formula.api as smf
import statsmodels.api as sm
import matplotlib.pyplot as plt
from scipy import stats
from sklearn.metrics import roc_curve, auc, classification_report, confusion_matrix

ces = pd.read_parquet("CES2020_clean.parquet")


# ==============================================================================
# SECTION 1: SIMPLE LOGIT — DID RESPONDENT VOTE?
# ==============================================================================

# smf.logit() uses the same formula syntax as smf.ols().
# family = Binomial with logit link is the default.
# fit() returns a LogitResults object with the same summary() interface.

m_voted_simple = smf.logit("voted ~ education", data=ces).fit()
print(m_voted_simple.summary())

# Coefficients are in LOG-ODDS units.
# Use np.exp() to convert to odds ratios.


# ==============================================================================
# SECTION 2: ODDS RATIOS
# ==============================================================================

def odds_ratio_table(result):
    """Return a DataFrame with ORs and 95% CIs."""
    coefs = result.params
    conf  = result.conf_int()
    pvals = result.pvalues

    or_df = pd.DataFrame({
        "OR":       np.exp(coefs),
        "CI_low":   np.exp(conf[0]),
        "CI_high":  np.exp(conf[1]),
        "p_value":  pvals,
    }).round(4)

    or_df["sig"] = or_df["p_value"].apply(
        lambda p: "***" if p < .001 else "**" if p < .01 else "*" if p < .05 else ""
    )
    return or_df

print("\nOdds Ratios — Simple Turnout Model:")
print(odds_ratio_table(m_voted_simple))


# ==============================================================================
# SECTION 3: FULL TURNOUT MODEL
# ==============================================================================

m_voted_full = smf.logit(
    "voted ~ education + age + C(sex) + C(census_region) "
    "+ ideology5 + party_id7 + college + white_nh + imm_restrict + econ_retro",
    data=ces
).fit(cov_type="HC3")

print(m_voted_full.summary())
print("\nORs — Full Turnout Model:")
print(odds_ratio_table(m_voted_full))


# ==============================================================================
# SECTION 4: BIDEN VOTE CHOICE (VOTERS ONLY)
# ==============================================================================

ces_voters = ces[(ces["voted"] == 1) & ces["biden_voter"].notna()].copy()

m_biden = smf.logit(
    "biden_voter ~ education + age + C(sex) + C(census_region) "
    "+ ideology5 + party_id7 + college + white_nh + imm_restrict + econ_retro",
    data=ces_voters
).fit(cov_type="HC3")

print("\nBiden Vote Choice — ORs:")
print(odds_ratio_table(m_biden))


# ==============================================================================
# SECTION 5: AVERAGE MARGINAL EFFECTS (AME)
# ==============================================================================

# statsmodels computes marginal effects via .get_margeff()
# at="overall" = average marginal effect (AME) — averaged across all observations
# at="mean"    = marginal effect at the mean (MEM)
# method="dydx" = derivative (change in probability per unit change in X)

m_base = smf.logit(
    "voted ~ education + age + C(sex) + ideology5 + party_id7",
    data=ces
).fit(disp=False)

me = m_base.get_margeff(at="overall", method="dydx")
print(me.summary())

# The "dy/dx" column shows: on average, a 1-unit increase in X is associated
# with this many percentage points change in Pr(voted=1).


# ==============================================================================
# SECTION 6: PREDICTED PROBABILITIES AT SPECIFIC VALUES
# ==============================================================================

# Build a prediction grid: vary education from 1-6, hold others at their mean
pred_grid = pd.DataFrame({
    "education":    range(1, 7),
    "age":          ces["age"].mean(),
    "sex":          1,
    "ideology5":    ces["ideology5"].mean(),
    "party_id7":    ces["party_id7"].mean(),
})

pred_grid["pr_voted"] = m_base.predict(pred_grid)

# Confidence intervals via delta method (manual)
# For clean CIs use the margins approach or bootstrap

print("\nPredicted Pr(voted) by education:")
print(pred_grid[["education", "pr_voted"]].round(3))

# Plot
plt.figure(figsize=(8, 5))
plt.plot(pred_grid["education"], pred_grid["pr_voted"],
         color="navy", linewidth=2, marker="o")
plt.xticks(range(1, 7),
           ["No HS", "HS grad", "Some coll.", "Assoc.", "Bach.", "Postgrad"])
plt.xlabel("Education Level")
plt.ylabel("Pr(Voted)")
plt.title("Predicted Probability of Voting by Education")
plt.ylim(0, 1)
plt.grid(alpha=0.3)
plt.tight_layout()
plt.savefig("pred_prob_voted_educ.png", dpi=300)
plt.close()


# ==============================================================================
# SECTION 7: PREDICTED PROBABILITY BY SEX AND EDUCATION
# ==============================================================================

from itertools import product

grid = pd.DataFrame(
    list(product(range(1, 7), [1, 2])),
    columns=["education", "sex"]
)
grid["age"]       = ces["age"].mean()
grid["ideology5"] = ces["ideology5"].mean()
grid["party_id7"] = ces["party_id7"].mean()

m_interact = smf.logit(
    "voted ~ education * C(sex) + age + ideology5",
    data=ces
).fit(disp=False)

grid["pr_voted"] = m_interact.predict(grid)

fig, ax = plt.subplots(figsize=(8, 5))
for sex_val, label, color in [(1, "Male", "navy"), (2, "Female", "firebrick")]:
    sub = grid[grid["sex"] == sex_val]
    ax.plot(sub["education"], sub["pr_voted"],
            color=color, label=label, linewidth=2, marker="o")

ax.set_xticks(range(1, 7))
ax.set_xticklabels(["No HS","HS","Some coll.","Assoc.","Bach.","Postgrad"])
ax.set_xlabel("Education Level")
ax.set_ylabel("Pr(Voted)")
ax.set_title("Predicted Pr(Voted) by Education and Sex")
ax.legend()
ax.set_ylim(0, 1)
plt.tight_layout()
plt.savefig("pred_prob_voted_sex.png", dpi=300)
plt.close()


# ==============================================================================
# SECTION 8: GOODNESS OF FIT
# ==============================================================================

# McFadden's pseudo R-squared
# = 1 - (log-likelihood of full / log-likelihood of null)

null_model = smf.logit("voted ~ 1", data=ces).fit(disp=False)
pseudo_r2  = 1 - (m_voted_full.llf / null_model.llf)
print(f"\nMcFadden pseudo R2: {pseudo_r2:.3f}")

# AIC and BIC
print(f"AIC: {m_voted_full.aic:.1f}")
print(f"BIC: {m_voted_full.bic:.1f}")

# Classification table at 0.5 threshold
y_true = m_voted_full.model.endog
y_pred = (m_voted_full.predict() > 0.5).astype(int)

print("\nClassification Report:")
print(classification_report(y_true, y_pred, target_names=["Did not vote", "Voted"]))

cm = confusion_matrix(y_true, y_pred)
print("Confusion matrix:\n", cm)

accuracy = (y_pred == y_true).mean()
print(f"Accuracy: {accuracy:.3f}")


# ==============================================================================
# SECTION 9: ROC CURVE AND AUC
# ==============================================================================

y_prob = m_voted_full.predict()
fpr, tpr, thresholds = roc_curve(y_true, y_prob)
roc_auc = auc(fpr, tpr)

print(f"\nAUC: {roc_auc:.3f}")
# AUC > 0.7 = acceptable; > 0.8 = excellent

plt.figure(figsize=(7, 6))
plt.plot(fpr, tpr, color="navy", linewidth=2,
         label=f"ROC Curve (AUC = {roc_auc:.3f})")
plt.plot([0, 1], [0, 1], color="gray", linestyle="--", label="Random (AUC = 0.5)")
plt.xlabel("False Positive Rate (1 - Specificity)")
plt.ylabel("True Positive Rate (Sensitivity)")
plt.title("ROC Curve: Voter Turnout Logit Model")
plt.legend()
plt.tight_layout()
plt.savefig("roc_turnout.png", dpi=300)
plt.close()


# ==============================================================================
# SECTION 10: PROBIT AS ALTERNATIVE
# ==============================================================================

m_probit = smf.probit(
    "voted ~ education + age + C(sex) + ideology5 + party_id7",
    data=ces
).fit(disp=False)

print(m_probit.summary())

# AMEs from probit — should be similar to logit AMEs
me_probit = m_probit.get_margeff(at="overall")
print(me_probit.summary())


# ==============================================================================
# SECTION 11: SIDE-BY-SIDE RESULTS TABLE
# ==============================================================================

l1 = smf.logit("voted ~ education + age + C(sex) + ideology5", data=ces).fit(disp=False)
l2 = smf.logit("voted ~ education + age + C(sex) + ideology5 + party_id7 + college", data=ces).fit(disp=False)
l3 = smf.logit("voted ~ education + age + C(sex) + ideology5 + party_id7 + college + white_nh + imm_restrict", data=ces).fit(disp=False)

for name, m in [("Model 1", l1), ("Model 2", l2), ("Model 3", l3)]:
    print(f"\n{name}  —  AIC={m.aic:.0f}  N={int(m.nobs)}")
    print(odds_ratio_table(m).to_string())

print("\nModule 4 complete.")

File: code-library/04-logistic-regression/logistic-regression.R
# ==============================================================================
#  MODULE 4: LOGISTIC REGRESSION
#  Dataset: CES 2020 (cleaned in Module 1)
#
#  Purpose: Predict binary outcomes (voted, biden_voter) using logistic
#  regression. Covers model estimation, odds ratios, average marginal effects,
#  goodness of fit, predicted probabilities, and ROC/AUC.
# ==============================================================================

library(tidyverse)
library(broom)
library(modelsummary)
library(sandwich)
library(lmtest)
library(marginaleffects)  # average marginal effects (install.packages("marginaleffects"))
library(pROC)             # ROC curves and AUC (install.packages("pROC"))

ces <- readRDS("CES2020_clean.rds")


# ==============================================================================
# SECTION 1: SIMPLE LOGIT — DID RESPONDENT VOTE?
# ==============================================================================

# glm(outcome ~ predictors, data, family = binomial(link = "logit"))
# family = binomial gives logistic regression
# family = binomial(link = "probit") gives probit
#
# Output coefficients are in LOG-ODDS units — not directly interpretable
# as probabilities. Use exp() for odds ratios, or margins for probability scale.

m_voted_simple <- glm(
  voted ~ education,
  data   = ces,
  family = binomial(link = "logit")
)

summary(m_voted_simple)
tidy(m_voted_simple, conf.int = TRUE)


# ==============================================================================
# SECTION 2: ODDS RATIOS
# ==============================================================================

# Exponentiate coefficients and confidence intervals to get odds ratios.
# exp(coef) > 1 = predictor increases odds of Y=1
# exp(coef) < 1 = predictor decreases odds of Y=1

tidy(m_voted_simple, exponentiate = TRUE, conf.int = TRUE)

# Example interpretation:
# OR for education = 1.35 means each additional unit of education
# multiplies the odds of voting by 1.35 (35% increase in odds).


# ==============================================================================
# SECTION 3: FULL TURNOUT MODEL
# ==============================================================================

m_voted_full <- glm(
  voted ~ education + age + factor(sex) + factor(census_region) +
    ideology5 + party_id7 + college + white_nh + imm_restrict + econ_retro,
  data   = ces,
  family = binomial(link = "logit")
)

summary(m_voted_full)

# Odds ratio table
tidy(m_voted_full, exponentiate = TRUE, conf.int = TRUE) |>
  mutate(across(where(is.numeric), ~ round(.x, 3))) |>
  print(n = 30)

# With robust SEs
coeftest(m_voted_full, vcov = vcovHC(m_voted_full, type = "HC3"))


# ==============================================================================
# SECTION 4: PREDICTING BIDEN VOTE CHOICE
# ==============================================================================

# Restrict to voters who chose between Biden and Trump
ces_voters <- ces |> filter(voted == 1, !is.na(biden_voter))

m_biden <- glm(
  biden_voter ~ education + age + factor(sex) + factor(census_region) +
    ideology5 + party_id7 + college + white_nh + imm_restrict + econ_retro,
  data   = ces_voters,
  family = binomial(link = "logit")
)

tidy(m_biden, exponentiate = TRUE, conf.int = TRUE) |>
  mutate(across(where(is.numeric), ~ round(.x, 3)))


# ==============================================================================
# SECTION 5: AVERAGE MARGINAL EFFECTS (AME)
# ==============================================================================

# Marginal effects convert log-odds to probability scale:
# dP/dX = average change in Pr(Y=1) for a 1-unit change in X,
#         averaged across all observations.
#
# marginaleffects::avg_slopes() computes AMEs for all predictors.
# This is what most applied social scientists report in place of log-odds.

ame_voted <- avg_slopes(m_voted_full)
ame_voted

# Round and display cleanly
ame_voted |>
  as_tibble() |>
  select(term, contrast, estimate, std.error, p.value, conf.low, conf.high) |>
  mutate(across(where(is.numeric), ~ round(.x, 4))) |>
  arrange(p.value)

# Interpretation: the estimate for education means "on average across
# respondents, a 1-unit increase in education is associated with a
# [estimate] change in Pr(voted=1)."


# ==============================================================================
# SECTION 6: PREDICTED PROBABILITIES AT SPECIFIC VALUES
# ==============================================================================

# predictions() computes Pr(Y=1) at specified covariate values.
# datagrid() creates a grid of combinations to evaluate at.

# Predicted Pr(voted) across education levels, holding others at mean
pred_educ <- predictions(
  m_voted_full,
  newdata = datagrid(
    education = 1:6,
    sex = 1,
    census_region = 1
  )
)

pred_educ |>
  as_tibble() |>
  select(education, estimate, conf.low, conf.high)

# Plot
pred_educ |>
  ggplot(aes(x = education, y = estimate)) +
  geom_line(color = "navy", linewidth = 1) +
  geom_ribbon(aes(ymin = conf.low, ymax = conf.high), alpha = 0.2, fill = "navy") +
  scale_x_continuous(
    breaks = 1:6,
    labels = c("No HS", "HS grad", "Some coll.", "Assoc.", "Bach.", "Postgrad")
  ) +
  labs(
    title  = "Predicted Probability of Voting by Education",
    x      = "Education Level",
    y      = "Pr(Voted)"
  ) +
  ylim(0, 1) +
  theme_minimal()


# ==============================================================================
# SECTION 7: PREDICTED PROBABILITY BY SEX AND EDUCATION (INTERACTION)
# ==============================================================================

m_interact <- glm(
  voted ~ education * factor(sex) + age + ideology5,
  data   = ces,
  family = binomial()
)

pred_sex_educ <- predictions(
  m_interact,
  newdata = datagrid(
    education = 1:6,
    sex       = 1:2
  )
)

ggplot(pred_sex_educ, aes(x = education, y = estimate,
                          color = factor(sex), fill = factor(sex))) +
  geom_line(linewidth = 1) +
  geom_ribbon(aes(ymin = conf.low, ymax = conf.high), alpha = 0.15) +
  scale_color_manual(values = c("navy", "firebrick"),
                     labels = c("Male", "Female")) +
  scale_fill_manual(values  = c("navy", "firebrick"),
                    labels = c("Male", "Female")) +
  scale_x_continuous(breaks = 1:6,
                     labels = c("No HS","HS","Some coll.","Assoc.","Bach.","Postgrad"),
                     guide  = guide_axis(angle = 45)) +
  labs(title  = "Predicted Pr(Voted) by Education and Sex",
       x      = "Education Level", y = "Pr(Voted)",
       color  = "Sex", fill = "Sex") +
  ylim(0, 1) +
  theme_minimal()


# ==============================================================================
# SECTION 8: GOODNESS OF FIT
# ==============================================================================

# McFadden's pseudo R-squared
# = 1 - (log-likelihood of full model / log-likelihood of intercept-only model)

null_model <- glm(voted ~ 1, data = ces, family = binomial())

pseudo_r2 <- 1 - (logLik(m_voted_full) / logLik(null_model))
cat("McFadden pseudo R2:", round(as.numeric(pseudo_r2), 3), "\n")

# Hosmer-Lemeshow test (from ResourceSelection package)
# install.packages("ResourceSelection")
library(ResourceSelection)

pred_probs <- predict(m_voted_full, type = "response")
hl_test    <- hoslem.test(m_voted_full$y, pred_probs, g = 10)
hl_test
# Non-significant p-value = adequate fit.

# Classification table at 0.5 threshold
ces_complete <- ces |>
  filter(!is.na(voted), !is.na(education), !is.na(age), !is.na(sex),
         !is.na(census_region), !is.na(ideology5), !is.na(party_id7),
         !is.na(college), !is.na(white_nh), !is.na(imm_restrict),
         !is.na(econ_retro))

pred_class <- if_else(fitted(m_voted_full) > 0.5, 1, 0)
table(Predicted = pred_class, Actual = m_voted_full$y)

accuracy <- mean(pred_class == m_voted_full$y)
cat("Accuracy:", round(accuracy, 3), "\n")


# ==============================================================================
# SECTION 9: ROC CURVE AND AUC
# ==============================================================================

roc_obj <- roc(m_voted_full$y, fitted(m_voted_full))

cat("AUC:", round(auc(roc_obj), 3), "\n")
# AUC > 0.7 = acceptable; > 0.8 = excellent; 0.5 = no better than chance

# Plot ROC curve using ggplot via pROC
ggroc(roc_obj, color = "navy", linewidth = 1) +
  geom_segment(aes(x = 1, xend = 0, y = 0, yend = 1),
               color = "gray50", linetype = "dashed") +
  labs(
    title    = paste("ROC Curve — AUC =", round(auc(roc_obj), 3)),
    x        = "Specificity",
    y        = "Sensitivity"
  ) +
  theme_minimal()


# ==============================================================================
# SECTION 10: PROBIT AS ALTERNATIVE
# ==============================================================================

m_probit <- glm(
  voted ~ education + age + factor(sex) + ideology5 + party_id7,
  data   = ces,
  family = binomial(link = "probit")
)

# AMEs from probit should be very similar to logit AMEs
avg_slopes(m_probit) |>
  as_tibble() |>
  select(term, estimate, std.error, p.value) |>
  mutate(across(where(is.numeric), ~ round(.x, 4)))


# ==============================================================================
# SECTION 11: PUBLICATION TABLE
# ==============================================================================

m_l1 <- glm(voted ~ education + age + factor(sex) + ideology5,
             data = ces, family = binomial())

m_l2 <- glm(voted ~ education + age + factor(sex) + ideology5 + party_id7 + college,
             data = ces, family = binomial())

m_l3 <- glm(voted ~ education + age + factor(sex) + ideology5 + party_id7 +
               college + white_nh + imm_restrict,
             data = ces, family = binomial())

# Logit table with ORs (exponentiate = TRUE)
modelsummary(
  list("Model 1" = m_l1, "Model 2" = m_l2, "Model 3" = m_l3),
  exponentiate = TRUE,
  stars   = c("*" = .05, "**" = .01, "***" = .001),
  gof_map = c("nobs", "logLik", "AIC"),
  title   = "Logistic Regression (Odds Ratios): Voter Turnout 2020"
)

message("Module 4 complete.")
